test/rate_limiter_test.cc

```diff
@@ -172,4 +172,102 @@ TEST_F(RateLimiterTest, DistributionSamplingRateLimiterImplSchedulingTest) {
   EXPECT_TRUE(rate_limiter->tryAcquireOne());
 }
 
+class LinearRampingRateLimiterImplTest : public Test {
+public:
+  std::vector<int64_t> getAcquisitionTimings(const Frequency frequency,
+                                             const std::chrono::seconds duration) {
+    Envoy::Event::SimulatedTimeSystem time_system;
+    std::vector<int64_t> aquisition_timings;
+    LinearRampingRateLimiterImpl rate_limiter(time_system, duration, frequency);
+    auto total_us_elapsed = 0us;
+    const auto clock_tick = 50us;
+    EXPECT_FALSE(rate_limiter.tryAcquireOne());
+    do {
+      if (rate_limiter.tryAcquireOne()) {
+        EXPECT_FALSE(rate_limiter.tryAcquireOne());
+        aquisition_timings.push_back(total_us_elapsed.count());
+      }
+      time_system.sleep(clock_tick);
+      total_us_elapsed += clock_tick;
+    } while (total_us_elapsed <= duration);
+    return aquisition_timings;
+  }
+};
+
+TEST_F(RateLimiterTest, LinearRampingRateLimiterImplInvalidArgumentTest) {
+  Envoy::Event::SimulatedTimeSystem time_system;
+  // bad frequency
+  EXPECT_THROW(LinearRampingRateLimiterImpl rate_limiter(time_system, 1s, 0_Hz);
+               , NighthawkException);
+  // bad ramp duration
+  EXPECT_THROW(LinearRampingRateLimiterImpl rate_limiter(time_system, 0s, 1_Hz);
+               , NighthawkException);
+  EXPECT_THROW(LinearRampingRateLimiterImpl rate_limiter(time_system, -1s, 1_Hz);
+               , NighthawkException);
+}
+
+TEST_F(LinearRampingRateLimiterImplTest, TimingVerificationTest) {
+  // The first acquisition timing (1000050) seems slightly off because of floating point
+  // rounding causing the first timing to fall into the next clock tick. This is
+  // a test only issue: in practice we don't have fixed microsecond step sizes,
+  // and the rate limiter computes at nanosecond precision internally. As we
+  // want to have microsecond level precision, this should be more then sufficient.
+  EXPECT_EQ(getAcquisitionTimings(5_Hz, 5s),
+            std::vector<int64_t>({1000050, 1732100, 2236100, 2645800, 3000000, 3316650, 3605600,
```

Comment:
Tests apart from testing are often times used by the readers to understand what the code does. (i.e. as additional documentation). Feels like the values that we chose for these tests don't really help the reader. Additionally the unit of these values isn't obvious since the last change to micro-seconds.

I am wondering if we could choose different parameters than 5Hz over 5sec or maybe add test cases with other parameters that will be easier / more obvious for the reader to comprehend. We don't want to decrease test coverage, but we should aim towards self-documenting and readable test cases. Can we try to find a balance?

Reply:
@mum4k I couldn't really obtain easy-to-grok numbers by picking different parameter values here; but how about this: 9d248f2a914e6cf667216ebac2a1ba46d5e5e618 ?

Instead of matching expectations against hard-coded numbers, that uses the second law of motion to figure out control acquisition timings for comparison, and adds a little slack to when verifying expectations with a comment explaining the need for that. I'm hoping that will help future readers of this code. wdyt, will this be better?

(As a side-effect, this approach also allowed adding tests with higher frequencies / duration arguments)